### ISSUE TITLE
1Password integration asks for MFA every time a credential is used

### ISSUE THREAD

**Issue by pv2b** [royalapplications/toolbox#54]
When using the 1Password integration, with an account with MFA enabled, every time we want to use a credential, I get an MFA prompt. That's not really ideal.

This was initially surprising, because 1Password is only supposed to require MFA when signing in on a new device, but investigating the code, it seems that the integration logs off manually every time. Getting a list of items is basically the sequence:

- Try to log off (ignore failures)
- Log on
- Get a list of items and convert it into Royal TS format
- Log off

The process is similar for a dynamic credential.

The result of this is that every login or reload of the dynamic folder is a new sign-in to the 1Password servers, which means a new MFA prompt. That's obviously less than ideal.

I tried, as a workaround, to set "Dynamic Credentials" to No, with the idea that it would just get all the passwords and store them in the Royal TS file, but for the amount of objects I have in 1Password it takes a very long time to load them all (after 15 minutes it still wasn't done so I gave up waiting), and I really prefer not storing my passwords in a Royal TS file if it can be avoided, I'd rather just keep them in the password manager.

Probably the best fix for this, in terms of functionality, would be to remove any logon/logoff logic from the Royal TS integration and expect the end user to directly interact with the op CLI for logging in and out. Alternatively, it should only try to log in if it can detect that it's not logged in, and not always be signing out. (I can imagine that would break some scenarios with multiple 1Password accounts in different files.)

I'm willing to put in some work into changing this, but I'd love to hear any feedback, maybe some insight into why the feature works as it does right now, any kind of edge cases I'm not thinking of, etc.

**Comment by pv2b** [comment]
I've done some further investigation on this, and the approach I outlined in the initial report won't work. 1Password sessions only last 30 minutes, and anyway you need to somehow persist the session key. Since there script is not designed to use any kind of persistent storage or state that's not really possible, so it makes the logic to "sign in" every time a little bit more logical. Therefore an approach where the user logs in themselves using the CLI will be impractical, they'd have to log in and then somehow populate a session key into the Royal TS GUI which then would only last 30 minutes, it'd be pretty terrible.

However, I did discover that 2FA prompts only happens when you *add* a new account. If you sign in to an *existing* acocunt, 2FA prompt does not happen. So, instead, I'm going to try an approach where the integration avoids re-creating an account if it already exists.

**Comment by lemonmojo** [comment]
Hey @pv2b!

First off, I totally agree that the experience is suboptimal when MFA is enabled. There's definitely room for improvement here.
However, as you have noted we took great care to design the scripts to be stateless and not leave any traces on the system they're executed on.

The second approach you're suggesting sounds good to me. However, I think it should be optional and configurable by the user. For non-MFA scenarios I think it's best to keep the current stateless design. So I suggest that a new setting is to be added in the custom properties to control that behavior which is off by default.

Please also note that we support both, V1 and V2 of the 1Password CLI and the script has to work properly on both, macOS and Windows.

If you can get that to work, we're happy to help with testing your changes and will merge once we're confident that everything works as expected.

Many thx!

cheers,
Felix

**Comment by pv2b** [comment]
Hello, again!

I hacked a bit on this and I think I've found a workable solution.

Basically, this extends the script to try to re-use an existing account instead of adding a new one every time. This avoids repeated MFA prompts but otherwise leaves the functionality unchanged.

If there's any error logging in with the saved account, or if the account doesn't exist, the script falls back on previous behaviour.

These behavioural changes only happen for 1Password Cli v2, if v1 is used behind the scenes this should not change its behaviour overall.

Just FYI, if your goal is to "leave no trace" on the system, you're not succeeding with that right now, because you're merely signing out from the account, not actually "forgetting" it. But because that's contrary to at least my goals, I'm not bothered by that at all. :-)

I will send in a seperate pull request for the new RDFE file, along with a DIFF showing the actual code so it's actually possible to read and review it.
